src/tfr_sensor/src/light_detector.cpp

```diff
@@ -0,0 +1,85 @@
+#include <light_detector.h>
+
+/*
+ *  add an image to the moving average, drops old images if no longer needed.
+ * */
+void LightDetector::add_image(const sensor_msgs::ImageConstPtr& msg)
+{
+    //manage the moving average
+    if (brightness.size() >= width)
+        brightness.pop_back();
+    
+    //convert out of std ros image
+    cv::Mat image;
+    try
+    {
+        image = cv_bridge::toCvCopy(msg)->image;
+    }
+    catch (cv_bridge::Exception& e)
+    {
+        ROS_ERROR("cv_bridge exception: %s",
+                e.what());
+        return;
+    }
+    cv::Scalar intensities = cv::sum(image);
+
+    ColorStats stats{};
+    auto scale = [&image] (const double &intensity) 
+    {
+        return intensity/(image.rows*image.cols);
+    };
+
+    //note we have to reverse out of native cv bgr ordering
+    stats.r_ave=scale(intensities[2]);
+    stats.g_ave=scale(intensities[1]);
+    stats.b_ave=scale(intensities[0]);
```

Comment:
In the future, let's make sure lambda expressions really make sense before we write them.  Most people, myself included, don't use them very often and are fairly unfamiliar with their syntax, which means it takes a long time to figure out what's going on.  Especially here, since it could have been as simple as `stats.r_ave = intensities[2] / (image.rows * image.cols);`.

Reply:
I would argue the lambda makes this more tedious calculation more readable. The name of the lambda tells you exactly what I am doing, scaling the image intensity. 

If I ever want to modify how I am scaling the image in the future, I just change the scaling function. Then it does all 3 channels at once. There is also like 0 extra performance overhead. 

Most of the modern 11, 14, and 17 standards make heavy use of them as a full blown replacements for functors, so lambdas are here to stay, and I think will become more and more prevalent as a useful construct.

Reply:
Replacements for functors, yes.  That is their purpose and they're a great way for generating functors.  You don't need functors here.

Reply:
(looked through all the cases where this happened before writing this response): I feel like this case could go either way in my opinion, although I do ask that if we stay with a lambda I think we might want a different name than "scale". For me "scale" implies resizing the image based on the work I've done in imaging programs, and given that we're working with images here that makes it more confusing (for me at least).

Reply:
We obviously still need to figure out our stance on the lambdas, but why are we changing from an array to a class for the colors? In my opinion, nothing would be more clear, concise and maintainable than a for loop.
for(int i =0;i < 3; i++)
    stats[i] = intensities[i]/image.rows*image.cols;